src/core/coap/coap.cpp

```diff
@@ -107,20 +107,27 @@ Message *CoapBase::NewMessage(const otMessageSettings *aSettings)
     return message;
 }
 
-otError CoapBase::SendMessage(Message &               aMessage,
-                              const Ip6::MessageInfo &aMessageInfo,
-                              otCoapResponseHandler   aHandler,
-                              void *                  aContext)
+otError CoapBase::SendMessage(Message &                           aMessage,
+                              const Ip6::MessageInfo &            aMessageInfo,
+                              otCoapResponseHandler               aHandler,
+                              void *                              aContext,
+                              const otCoapTransmissionParameters *aTransmissionParameters)
 {
-    otError      error;
-    CoapMetadata coapMetadata;
-    Message *    storedCopy = NULL;
-    uint16_t     copyLength = 0;
+    otError  error;
+    Message *storedCopy = NULL;
+    uint16_t copyLength = 0;
+
+    if (aTransmissionParameters == NULL)
+    {
+        aTransmissionParameters = &mDefaultTransmissionParameters;
+    }
+    const CoapTransmissionParameters &transmissionParameters =
+        *static_cast<const CoapTransmissionParameters *>(aTransmissionParameters);
```

Comment:
I think elsewhere in the code, this `static_cast` would be done in `src/core/api/coap_api.cpp`, and the C++ routine would have a `ot::Coap::CoapTransmissionParameters&` parameter.

I wonder if this should be moved for consistency with existing code (e.g. look at where `otMessage` gets cast to `ot::Coap::Message`).

Reply:
Hi @sjlongland, thanks for reviewing.

Originally I had the same thoughts, but then I found instances where it was done otherwise, for example:
https://github.com/openthread/openthread/blob/73ff9439e60c22735f4e9a7f1e6b040a7d3860fd/src/core/net/ip6.cpp#L75-L85

@sjlongland do you think it would be neater to cast to CoapTransmissionParameters& "upstream"?

Reply:
I prefer to have the cast in `src/core/api` since that is the boundary between C and C++.

Note that the cited `Ip6::NewMessage()` example above was recently removed as a part of #4522.

Reply:
Implemented cast in `src/core/api` as suggested by @sjlongland, @jwhui and @abtink (btw thanks all for reviewing).